Add forward AD formulas for `{adaptive_,fractional_,}max_pool{2,3}d_{backward,}`

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70355
* #69956
* #69998
* __->__ #69884

Also fixes: https://github.com/pytorch/pytorch/issues/69322, https://github.com/pytorch/pytorch/issues/69325

Differential Revision: [D33093039](https://our.internmc.facebook.com/intern/diff/D33093039)

soulitzer committed:
Add forward AD formulas for {adaptive_,fractional_,}max_pool{2,3}d_{backward,}

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for `{adaptive_,fractional_,}max_pool{2,3}d_{backward,}`"


Also fixes: https://github.com/pytorch/pytorch/issues/69322, https://github.com/pytorch/pytorch/issues/69325


[ghstack-poisoned]

albanD reviewed (commented):
Code change looks ok. Let's see what CI thinks.

soulitzer commented on tools/autograd/derivatives.yaml:
```diff
@@ -1951,9 +1951,13 @@
 - name: adaptive_max_pool2d(Tensor self, int[2] output_size) -> (Tensor, Tensor)
   self: adaptive_max_pool2d_backward(grad, self, result1)
+  result0: gather(self_t.flatten(-2), -1, result1.flatten(-2)).view_as(result1)
+  output_differentiability: [True, False]
```
Hmm maybe, codegen doesn't seem to complain when I remove it, but maybe we should keep it just so that it is more explicit?

albanD replied:
I'm sure that would be part of the differentiable/not implemented cleanup :)

soulitzer committed:
Update on "Add forward AD formulas for `{adaptive_,fractional_,}max_pool{2,3}d_{backward,}`"


Also fixes: https://github.com/pytorch/pytorch/issues/69322, https://github.com/pytorch/pytorch/issues/69325


[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for `{adaptive_,fractional_,}max_pool{2,3}d_{backward,}`"


Also fixes: https://github.com/pytorch/pytorch/issues/69322, https://github.com/pytorch/pytorch/issues/69325


[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for `{adaptive_,fractional_,}max_pool{2,3}d_{backward,}`"


Also fixes: https://github.com/pytorch/pytorch/issues/69322, https://github.com/pytorch/pytorch/issues/69325

Differential Revision: [D33093039](https://our.internmc.facebook.com/intern/diff/D33093039)


[ghstack-poisoned]

albanD reviewed (commented):
some lint issues are left?

soulitzer committed:
Update on "Add forward AD formulas for `{adaptive_,fractional_,}max_pool{2,3}d_{backward,}`"


Also fixes: https://github.com/pytorch/pytorch/issues/69322, https://github.com/pytorch/pytorch/issues/69325

Differential Revision: [D33093039](https://our.internmc.facebook.com/intern/diff/D33093039)


[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for `{adaptive_,fractional_,}max_pool{2,3}d_{backward,}`"


Also fixes: https://github.com/pytorch/pytorch/issues/69322, https://github.com/pytorch/pytorch/issues/69325

Differential Revision: [D33093039](https://our.internmc.facebook.com/intern/diff/D33093039)


[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for `{adaptive_,fractional_,}max_pool{2,3}d_{backward,}`"


Also fixes: https://github.com/pytorch/pytorch/issues/69322, https://github.com/pytorch/pytorch/issues/69325

Differential Revision: [D33093039](https://our.internmc.facebook.com/intern/diff/D33093039)


[ghstack-poisoned]

albanD reviewed (approved):
SGTM

soulitzer committed:
Update on "Add forward AD formulas for `{adaptive_,fractional_,}max_pool{2,3}d_{backward,}`"


Also fixes: https://github.com/pytorch/pytorch/issues/69322, https://github.com/pytorch/pytorch/issues/69325

Differential Revision: [D33093039](https://our.internmc.facebook.com/intern/diff/D33093039)


[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for `{adaptive_,fractional_,}max_pool{2,3}d_{backward,}`"


Also fixes: https://github.com/pytorch/pytorch/issues/69322, https://github.com/pytorch/pytorch/issues/69325

Differential Revision: [D33093039](https://our.internmc.facebook.com/intern/diff/D33093039)


[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33093039).

soulitzer committed:
Update on "Add forward AD formulas for `{adaptive_,fractional_,}max_pool{2,3}d_{backward,}`"


Also fixes: https://github.com/pytorch/pytorch/issues/69322, https://github.com/pytorch/pytorch/issues/69325

Differential Revision: [D33093039](https://our.internmc.facebook.com/intern/diff/D33093039)


[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33093039).

soulitzer committed:
Update on "Add forward AD formulas for `{adaptive_,fractional_,}max_pool{2,3}d_{backward,}`"


Also fixes: https://github.com/pytorch/pytorch/issues/69322, https://github.com/pytorch/pytorch/issues/69325

Differential Revision: [D33093039](https://our.internmc.facebook.com/intern/diff/D33093039)


[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33093039).
